### ISSUE TITLE
"Too much memory when copying: NNN"

### ISSUE THREAD

**Issue by parkr** [rbspy/rbspy#111]
Hi @jvns! I was trying to profile `jekyll build` on the latest master for a "massive" site (10,000 posts) and am running into an interesting memory issue! Have you seen this before?

I'm running on a Linode VM with 1GB memory. Do I need more memory? `jekyll build` runs fine. 😄 

```text
~/rbspy$ git rev-parse HEAD
a8e3c31bba61bec0c47269f9b39a1e5e7eb7a4c1
~/massive-jekyll-site$ free -m
              total        used        free      shared  buff/cache   available
Mem:            979         191         434          11         353         757
Swap:           511          47         464
~/massive-jekyll-site$ RUST_BACKTRACE=1 ~/rbspy/target/debug/rbspy record -- /home/parker/jekyll/exe/jekyll build
Press Ctrl+C to stop
Dropped 21/21 stack traces because of errors.
Error. Causes:
- Too much memory requested when copying: 2492672
stack backtrace:
   0:     0x560d772c50fc - backtrace::backtrace::libunwind::trace
                        at /home/parker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.5/src/backtrace/libunwind.rs:53
                         - backtrace::backtrace::trace<closure>
                        at /home/parker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.5/src/backtrace/mod.rs:42
   1:     0x560d772beeaf - backtrace::capture::{{impl}}::new_unresolved
                        at /home/parker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.5/src/capture.rs:88
   2:     0x560d772bdaf1 - failure::backtrace::internal::{{impl}}::new
                        at /home/parker/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/backtrace/internal.rs:44
   3:     0x560d772bb97e - failure::backtrace::{{impl}}::new
                        at /home/parker/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/backtrace/mod.rs:98
   4:     0x560d76d5e4f0 - failure::error::{{impl}}::from<rbspy::core::copy::MemoryCopyError>
                        at /home/parker/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/error.rs:34
   5:     0x560d76d0e1f7 - core::convert::{{impl}}::into<rbspy::core::copy::MemoryCopyError,failure::error::Error>
                        at /checkout/src/libcore/convert.rs:415
   6:     0x560d76de2f0f - rbspy::record
                        at src/main.rs:319
   7:     0x560d76de0bf6 - rbspy::do_main
                        at src/main.rs:161
   8:     0x560d76de1556 - rbspy::main
                        at src/main.rs:188
   9:     0x560d77308c9e - panic_unwind::__rust_maybe_catch_panic
                        at /checkout/src/libpanic_unwind/lib.rs:101
  10:     0x560d77300bbb - std::panicking::try<(),closure>
                        at /checkout/src/libstd/panicking.rs:459
                         - std::panic::catch_unwind<closure,()>
                        at /checkout/src/libstd/panic.rs:365
                         - std::rt::lang_start
                        at /checkout/src/libstd/rt.rs:58
  11:     0x560d76de8c3d - main
  12:     0x7f8bc2c082b0 - __libc_start_main
  13:     0x560d76cb9a49 - _start
  14:                0x0 - <unknown>
```

Thank you!!!!!!!!

**Comment by jvns** [comment]
thanks! this is a bug that got introduced recently that I actually just noticed today. going to work on fixing this.

**Comment by parkr** [comment]
Thank you so much, @jvns!

I just did a `git bisect` and 26c7594289820c046f525037418b3d9c64ad2f9c is the first bad commit.

**Comment by jvns** [comment]
weird, that commit only adds comments!

**Comment by parkr** [comment]
Yeah! I must have done it wrong...

**Comment by parkr** [comment]
Just did another bisect! This time, I got 8ceded9b2a076bcf7e7d9d027cfc5b05406694c0.

```text
8ceded9b2a076bcf7e7d9d027cfc5b05406694c0 is the first bad commit
commit 8ceded9b2a076bcf7e7d9d027cfc5b05406694c0
Author: Julia Evans <julia@jvns.ca>
Date:   Fri Feb 9 19:22:22 2018 -0500

    Say that duration is in seconds. Fixes #94.

:040000 040000 946a3b0182a4a57f03c686dc810bb50f6ff665dd c4623089dcfd15235429d0194516af4afc74d834 M      src
```

Using the following Makefile:

```makefile
all: rev compile jekyll

rev:
        git rev-parse HEAD

compile:
        cargo build

jekyll: compile
        ./target/debug/rbspy record -- \
          /home/parker/jekyll/exe/jekyll \
          build \
          -s /home/parker/massive-jekyll-site \
          -d /home/parker/massive-jekyll-site/_site
```

And running the following command in shell:

```console
$ (make && (git checkout Cargo.lock ; git bisect good)) || (git checkout Cargo.lock ; git bisect bad)
```

**Comment by parkr** [comment]
Well, I went back to v0.1.0 and got a third, different commit! Ugh. I must be doing something wrong :( 

This time it's not a documentation commit, though! 31f750a3b31ba7daa548a2b30489c873587b0655

```text
31f750a3b31ba7daa548a2b30489c873587b0655 is the first bad commit
commit 31f750a3b31ba7daa548a2b30489c873587b0655
Author: Julia Evans <julia@jvns.ca>
Date:   Fri Feb 9 14:21:14 2018 -0500

    ruby_version: Simplify stack base code

    This adds 2 new methods: `stack_field` and `stack_size_field` which just
    return t.stack / t.vm_stack and t.stack_size/t.vm_stack_size
    respectively, since those fields were renamed from Ruby 2.4 / 2.5 but
    have the same meaning.

    This makes it easier to write a bunch of the logic for the
    is_maybe_thread heuristic and the stack_base function (since it's all
    the same across Ruby versions except the names of those 2 fields)

:040000 040000 c4623089dcfd15235429d0194516af4afc74d834 4c8ff113ae18efd378cece5e21533b937affb2e6 M      src
```
